table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveMetaStoreRetryClient.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.ColumnStatisticsObj;
+import org.apache.hadoop.hive.metastore.api.InvalidInputException;
+import org.apache.hadoop.hive.metastore.api.InvalidObjectException;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A wrapper around HMS client that does caching and reconnection.
+ */
+public class HiveMetaStoreRetryClient implements MetaStoreClient {
+  private HiveMetaStoreClient mHive = null;
+  private String mConnectionUri;
+  private String mHiveDbName;
+
+  /**
+   * Constructor for the HiveMetaStoreRetryClient.
+   * @param connectionUri connection uri to the metastore
+   * @param hiveDbName the db name to access
+   */
+  public HiveMetaStoreRetryClient(String connectionUri, String hiveDbName) {
+    mConnectionUri = connectionUri;
+    mHiveDbName = hiveDbName;
+  }
+
+  private HiveMetaStoreClient getHive() throws IOException {
```

Comment:
I think we are going to have to make this thread safe, by using a lock to protect changes to the cached client. Probably a RW lock, read to get the client, write to update it.